pkg/logs/internal/framer/framer.go

```diff
@@ -0,0 +1,184 @@
+	case UTF8Newline:
+		matcher = &oneByteNewLineMatcher{contentLenLimit}
+	case UTF16BENewline:
+		matcher = &twoByteNewLineMatcher{contentLenLimit: contentLenLimit, newline: Utf16beEOL}
+	case UTF16LENewline:
+		matcher = &twoByteNewLineMatcher{contentLenLimit: contentLenLimit, newline: Utf16leEOL}
+	case SHIFTJISNewline:
+		// No special handling required for the newline matcher since Shift JIS does not use
+		// newline characters (0x0a) as the second byte of a multibyte sequence.
+		matcher = &oneByteNewLineMatcher{contentLenLimit}
+	case DockerStream:
+		matcher = &dockerStreamMatcher{contentLenLimit}
+	default:
+		panic(fmt.Sprintf("unknown framing %d", framing))
+	}
+
+	return &Framer{
+		frames:          0,
+		outputFn:        outputFn,
+		matcher:         matcher,
+		buffer:          bytes.Buffer{},
+		bytesFramed:     0,
+		contentLenLimit: contentLenLimit,
+	}
+}
+
+// GetFrameCount gets the number of frames this framer has processed.  This is safe to
+// call from any goroutine.
+func (fr *Framer) GetFrameCount() int64 {
+	return atomic.LoadInt64(&fr.frames)
+}
+
+// Process handles an incoming chunk of data.  It will call outputFn for any recognized frames.  Partial
+// frames are maintained between calls to Process.  The passed buffer is not used after return.
+func (fr *Framer) Process(inBuf []byte) {
+	// buffer is laid out as follows:
+	//
+	//                  /------from inBuf------\
+	// xxxxxxxxFFFFFFFFFFFFFFffffffffFFFFFFFffff
+	//  framed ^   seen ^
+	//
+	// Here "xx" is data that has already been sent to outputFn; and the "ff" and "FF" are
+	// as-yet un-recognized frames of data.  The `seen` offset indicates where the matcher
+	// left off in the last call to Process.
+
+	framed := fr.bytesFramed
+	seen := fr.buffer.Len()
+	fr.buffer.Write(inBuf)
+	end := fr.buffer.Len()
+	contentLenLimit := fr.contentLenLimit
+
+	for {
+		if framed == end {
+			break
+		}
+		buf := fr.buffer.Bytes()[framed:]
+
+		content, rawDataLen := fr.matcher.FindFrame(buf, seen-framed)
+		if content == nil {
+			// if the matcher was asked to match more than contentLenLimit,
+			// chop off contentLenLimit raw bytes and output them
+			if len(buf) >= contentLenLimit {
+				content, rawDataLen = buf[:contentLenLimit], contentLenLimit
+			} else {
+				// matcher didn't find a frame, so leave the remainder in
+				// buffer
+				break
+			}
+		}
+
+		// copy the data so that we can reuse the buffer
+		owned := make([]byte, len(content))
+		copy(owned, content)
+
+		fr.outputFn(owned, rawDataLen)
+		framed += rawDataLen
+		seen = framed
+	}
+
+	// check whether the buffer is empty, or if we should move the un-framed
+	// data to re-use the buffer space
+	if framed == fr.buffer.Len() {
+		fr.buffer.Reset()
+		framed = 0
+	} else if framed*2 > fr.buffer.Len() {
+		buf := fr.buffer.Bytes()
+		unframed := buf[framed:]
+		copy(buf[:len(unframed)], unframed)
+		fr.buffer.Truncate(len(unframed))
+		framed = 0
+	}
+	fr.bytesFramed = framed
+}
```

Comment:
Not a strong opinion, you may even dislike it, but what do you think of:

```
	// we've processed everything
	if framed == fr.buffer.Len() {
		fr.Reset()
		return
	}

	// check whether we should move the un-framed data to re-use the buffer space
 	if framed*2 > fr.buffer.Len() {
		buf := fr.buffer.Bytes()
		unframed := buf[framed:]
		copy(buf[:len(unframed)], unframed)
		fr.buffer.Truncate(len(unframed))
		fr.bytesFramed = 0
		return
	}

	// otherwise, move forward
	fr.bytesFramed = framed
```